cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java

```diff
@@ -111,10 +101,30 @@ public KafkaConsumerMetaTable createKafkaConsumerMeta() {
     }
   }
 
-  private MetricsTable getOrCreateMetricsTable(String tableName, DatasetProperties props)
-    throws DatasetManagementException, IOException {
+  private MetricsTable getOrCreateMetricsTable(String tableName, DatasetProperties props) {
+
+    MetricsTable table = null;
+    while (table == null) {
+      try {
+        table = DatasetsUtil.getOrCreateDataset(dsFramework, tableName,
+                                                MetricsTable.class.getName(), props, null, null);
+      } catch (DatasetManagementException e) {
+        // dataset service may be not up yet
+        // todo: seems like this logic applies everywhere, so should we move it to DatasetsUtil?
```

Comment:
not at this point: we want to keep retrying until we can write data. If we had, then options are to shut down or skip some data. None make sense really